src/coreclr/vm/weakreferencenative.cpp

```diff
@@ -10,1031 +10,173 @@
-        GCX_NOTRIGGER();
-
-        // If the existing handle is a GC weak handle and the new target is not an RCW, then
-        // we can avoid setting up a helper method frame and just reset the handle directly.
-        if (!IsNativeComWeakReferenceHandle(handle))
-        {
-            if (pTarget == NULL || !pTarget->GetMethodTable()->IsComObjectType())
-            {
-                StoreObjectInHandle(handle, pTarget);
-                storedObject = true;
-            }
-        }
-
-        // SetWeakReferenceTarget will reacquire the spinlock after setting up a helper method frame.  This allows
-        // the frame setup to throw without worrying about leaking the spinlock, and allows the epilog to be cleanly
-        // walked by the epilog decoder.
-        ReleaseWeakHandleSpinLock(pThis, handle);
-    }
-
-    // If we reset the handle directly, then early out before setting up a helper method frame
-    if (storedObject)
-    {
-        FC_GC_POLL();
-        return;
-    }
-
-    FC_INNER_RETURN_VOID(SetWeakReferenceTarget(pThis, pTarget, GetEEFuncEntryPointMacro(WeakReferenceOfTNative::SetTarget)));
-}
-FCIMPLEND
-
-#include <optdefault.h>
-
-//************************************************************************
-
-FCIMPL1(FC_BOOL_RET, WeakReferenceNative::IsTrackResurrection, WeakReferenceObject * pThisUNSAFE)
-{
-    FCALL_CONTRACT;
-
-    WEAKREFERENCEREF pThis(pThisUNSAFE);
-
-    if (pThis == NULL)
-    {
-        FCUnique(0x1);
-        FCThrow(kNullReferenceException);
-    }
-
-    BOOL trackResurrection = FALSE;
-    OBJECTHANDLE handle = AcquireWeakHandleSpinLock(pThis);
-    {
-        GCX_NOTRIGGER();
-
-        if (handle == NULL)
-        {
-            trackResurrection = FALSE;
-        }
-        else
-        if (IS_SPECIAL_HANDLE(handle))
-        {
-            trackResurrection = (handle == SPECIAL_HANDLE_FINALIZED_LONG);
-        }
-        else
-        {
-            trackResurrection = GCHandleUtilities::GetGCHandleManager()->HandleFetchType(GetHandleValue(handle)) == HNDTYPE_WEAK_LONG;
-        }
-
-        ReleaseWeakHandleSpinLock(pThis, handle);
-    }
-
-    FC_GC_POLL_RET();
-    FC_RETURN_BOOL(trackResurrection);
-}
-FCIMPLEND
-
-FCIMPL1(FC_BOOL_RET, WeakReferenceOfTNative::IsTrackResurrection, WeakReferenceObject * pThisUNSAFE)
+FCIMPL2(IWeakReference*, ComAwareWeakReferenceNative::ObjectToComWeakRef, Object* pObject, INT64* pWrapperId)
 {
     FCALL_CONTRACT;
+    _ASSERTE(pObject != nullptr);
 
-    WEAKREFERENCEREF pThis(pThisUNSAFE);
-
-    if (pThis == NULL)
+    SyncBlock* pSyncBlock = pObject->PassiveGetSyncBlock();
+    if (pSyncBlock != nullptr && pSyncBlock->GetInteropInfoNoCreate() != nullptr)
     {
-        FCThrow(kNullReferenceException);
+        FC_INNER_RETURN(IWeakReference*, ObjectToComWeakRefFramed(pObject, pWrapperId, GetEEFuncEntryPointMacro(ComAwareWeakReferenceNative::ObjectToComWeakRef)));
```

Comment:
I would rather split this into a FCall without HelperMethodFrame for the fast path, and regular QCall for the heavy lifting. (One of my very long-term goals is to get rid of all FCalls with HMF.)

Reply:
I actually considered making this check entirely in managed code. I already check for SyncBlock presence, so could check for interop info as well. I had concerns about whether that would be too much looking into native details from the managed code or whether that is safe if GC happens. 

I will take a look.

Reply:
> I actually considered making the fast path check entirely in managed code.

I think it would be pushing too much sync block details to managed code. I do not think we want to do that at this point.